src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
 }
 
+#endif /* OPENSSL_ALL */
 
-/* return code compliant with OpenSSL :
- *   size of shared secret if success, -1 if error
+#endif /* !NO_CERTS */
+
+#endif /* OPENSSL_EXTRA */
+
+#if defined(OPENSSL_EXTRA) ||  \
+ ((!defined(NO_BIO) || !defined(NO_FILESYSTEM)) && \
+  defined(HAVE_LIGHTY) || defined(HAVE_STUNNEL) || \
+  defined(WOLFSSL_MYSQL_COMPATIBLE))
+
+/* Load the DER encoded DH parameters into DH key.
+ *
+ * @param [in, out] dh      DH key to load parameters into.
+ * @param [in]      derBuf  Buffer holding DER encoded parameters data.
+ * @param [in]      derSz   Size of DER data in buffer in bytes.
+ *
+ * @return  1 on success.
+ * @return  -1 when DH or derBuf is NULL,
+ *                  internal DH key in DH is NULL,
+ *                  derSz is 0 or less,
+ *                  error decoding DER data or
+ *                  setting external parameter values fails.
  */
-int wolfSSL_DH_compute_key(unsigned char* key, const WOLFSSL_BIGNUM* otherPub,
-                          WOLFSSL_DH* dh)
+int wolfSSL_DH_LoadDer(WOLFSSL_DH* dh, const unsigned char* derBuf, int derSz)
 {
-    int            ret    = -1;
-    word32         keySz  = 0;
-    int            pubSz  = 1024;
-    int            privSz = 1024;
-#ifdef WOLFSSL_SMALL_STACK
-    unsigned char* pub;
-    unsigned char* priv   = NULL;
+    int    ret = 1;
+#if !defined(HAVE_FIPS) || FIPS_VERSION_GT(2,0)
+    word32 idx = 0;
 #else
-    unsigned char  pub [1024];
-    unsigned char  priv[1024];
+    byte* p = NULL;
+    byte* g = NULL;
+    word32 pSz = MAX_DH_SIZE;
+    word32 gSz = MAX_DH_SIZE;
 #endif
 
-    WOLFSSL_MSG("wolfSSL_DH_compute_key");
+    /* Validate parameters. */
+    if ((dh == NULL) || (dh->internal == NULL) || (derBuf == NULL) ||
+            (derSz <= 0)) {
+        WOLFSSL_ERROR_MSG("Bad function arguments");
+        ret = -1;
+    }
 
-#ifdef WOLFSSL_SMALL_STACK
-    pub = (unsigned char*)XMALLOC(pubSz, NULL, DYNAMIC_TYPE_PUBLIC_KEY);
-    if (pub == NULL)
-        return ret;
+#if !defined(HAVE_FIPS) || FIPS_VERSION_GT(2,0)
+    /* Decode the parameters/key into the internal DH key. */
+    if ((ret == 1) && (wc_DhKeyDecode(derBuf, &idx, (DhKey*)dh->internal,
+            (word32)derSz) < 0)) {
+        WOLFSSL_ERROR_MSG("wc_DhKeyDecode failed");
+        ret = -1;
+    }
+    if (ret == 1) {
+        /* Internal DH key has parameters. */
+        dh->inSet = 1;
 
-    priv = (unsigned char*)XMALLOC(privSz, NULL, DYNAMIC_TYPE_PRIVATE_KEY);
-    if (priv == NULL) {
-        XFREE(pub, NULL, DYNAMIC_TYPE_PUBLIC_KEY);
-        return ret;
+        /* Transfer parameters from internal to external DH key. */
+        if (SetDhExternal(dh) != 1) {
+            WOLFSSL_ERROR_MSG("SetDhExternal failed");
+            ret = -1;
+        }
+    }
+#else
+    /* Only DH parameters supported. */
+    /* Load external and set internal. */
+    if (ret == 1) {
+        /* Load data in manually */
+        p = (byte*)XMALLOC(pSz, NULL, DYNAMIC_TYPE_PUBLIC_KEY);
```

Comment:
Do p and g need to be freed in error cases? Seems like a leak is possible.

Reply:
Code doesn't work - removed.